rllib/evaluation/worker_set.py

```diff
@@ -601,49 +671,137 @@ def is_policy_to_train(
+    def foreach_worker(
+        self,
+        func: Callable[[RolloutWorker], T],
+        *,
+        local_worker=True,
+        # TODO(jungong) : switch to True once Algorithm is migrated.
+        healthy_only=False,
+        remote_worker_indices: List[int] = None,
+        timeout_seconds=None,
+    ) -> List[T]:
+        """Calls the given function with each worker instance as the argument.
 
         Args:
             func: The function to call for each worker (as only arg).
+            local_worker: Whether apply func on local worker too. Default is True.
+            healthy_only: Apply func on known active workers only. By default
+                this will apply func on all workers regardless of their states.
+            remote_worker_indices: Apply func on a selected set of remote workers.
+            timeout_seconds: Time to wait for results. Default is None.
 
         Returns:
              The list of return values of all calls to `func([worker])`.
         """
         local_result = []
-        if self.local_worker() is not None:
+        if local_worker and self.local_worker() is not None:
             local_result = [func(self.local_worker())]
-        remote_results = ray.get([w.apply.remote(func) for w in self.remote_workers()])
+
+        remote_results = self.__worker_manager.foreach_actor(
+            func,
+            healthy_only=healthy_only,
+            remote_actor_ids=remote_worker_indices,
+            timeout_seconds=timeout_seconds,
+        )
+        remote_results = [r.get() for r in remote_results.ignore_ray_errors()]
+
         return local_result + remote_results
 
     @DeveloperAPI
-    def foreach_worker_with_index(
-        self, func: Callable[[RolloutWorker, int], T]
+    def foreach_worker_with_id(
+        self,
+        func: Callable[[int, RolloutWorker], T],
+        *,
+        local_worker=True,
+        # TODO(jungong) : switch to True once Algorithm is migrated.
+        healthy_only=False,
+        remote_worker_indices: List[int] = None,
+        timeout_seconds=None,
     ) -> List[T]:
-        """Calls `func` with each worker instance and worker idx as args.
-
-        The index will be passed as the second arg to the given function.
+        """Similar to foreach_worker(), but calls the function with id of the worker too.
 
         Args:
-            func: The function to call for each worker and its index
-                (as args). The local worker has index 0, all remote workers
-                have indices > 0.
+            func: The function to call for each worker (as only arg).
+            local_worker: Whether apply func on local worker too. Default is True.
+            healthy_only: Apply func on known active workers only. By default
+                this will apply func on all workers regardless of their states.
+            remote_worker_indices: Apply func on a selected set of remote workers.
```

Comment:
wait, what is the range of acceptable remote_worker_indices? 0, ..., n-1  or 1, ..., n?

Reply:
depending on which init_id we use.
I added an API ``healthy_actor_ids()`` for the caller to get a valid list of ids to call.
I am using it in some of the algorithms.